MH-12758: Fix skipped AssetManagerDecorators

mtneug opened the pull request:
PR #103 introduced the new `takeSnapshot(MediaPackage mediaPackage)` method which was not overriden in the `AssetManagerDecorator` sub classes. The decorators were thus effectively skipped completely. This PR adds an override to the two sub decorators.

smarquard commented:
Apologies for missing this in the review of PR #103.

To clarify what's going on here, is it that when the ACL is changed in the UI, an update message is not sent through ActiveMQ so the change is not reflected in for example the Admin UI index?

mtneug commented:
Yes, that is the problem.

smarquard commented:
I can't reproduce the error that this is supposed to fix testing on stable.opencast.org - event ACLs appear to be saved and update correctly.

mtneug commented:
Yes this is not in 4.1 but in the current 4.x branch. #103 was not part of 4.1.

smarquard commented:
stable.opencast.org claims to be commit d9b26da which is after #103. Also I can't reproduce on our 4.x dev system which definitely includes #103.

lkiesow commented:
stable.opencast org is not 4.1 but is build from 4.x

mtneug commented:
OK my bad.

Could you test with an already recorded episode?

lkiesow commented:
Define recorded episode.

I can confirm that changing ACLs for existing episodes works on stabel.opencast.org.

mtneug commented:
I just uploaded "Test MH-12758 mtneug". Can you reproduce the following behaviour:

1. Go to Access Policy and look at the current ACLs. There should only be a policy allowing read and write access to ROLE_ADMIN
2. Now add e.g. ROLE_USER with read access. The Admin UI says that the ACLs are updated successfully.
3. Reload the page. ROLE_USER should be missing, because the Admin UI index was not updated.

lkiesow commented:
Hi @mtneug, I confirmed the problem and confirmed that your fix works.
However, would you mind going with this instead:

```diff
diff --git a/modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerDecorator.java b/modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerDecorator.java
index 7f2834f2e..61d1eb4c0 100644
--- a/modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerDecorator.java
+++ b/modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AssetManagerDecorator.java
@@ -39,11 +39,11 @@ public class AssetManagerDecorator implements AssetManager {
   }
 
   @Override public Snapshot takeSnapshot(String owner, MediaPackage mp) {
-    return delegate.takeSnapshot(owner, mp);
+    return owner == null ? delegate.takeSnapshot(mp) : delegate.takeSnapshot(owner, mp);
   }
 
   @Override public Snapshot takeSnapshot(MediaPackage mp) {
-    return delegate.takeSnapshot(mp);
+    return takeSnapshot(null, mp);
   }
 
   @Override public Opt<Asset> getAsset(Version version, String mpId, String mpeId) {
```

This avoids doing the exact same thing in every child.

mtneug committed:
MH-12758: Fix skipped AssetManagerDecorators

mtneug commented:
I updated the code as you suggested, although I think it is somewhat confusing that `owner` can be `null` in the decorator classes but not in the `AbstractAssetManager`. However, this is probably better than changing every child class.

lkiesow commented:
We could also change the AbstractAssetManager in the same way, letting the `takeSnapshot(String, MediaPackage)` itself fall back on the last owner if it's not set.

mtneug commented:
You mean something like that:
```diff
diff --git a/modules/matterhorn-asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AbstractAssetManager.java b/modules/matterhorn-asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AbstractAssetManager.java
index 85ce7ca40..df1a8ad25 100644
--- a/modules/matterhorn-asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AbstractAssetManager.java
+++ b/modules/matterhorn-asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AbstractAssetManager.java
@@ -54,7 +54,6 @@ import org.opencastproject.util.Checksum;
 import org.opencastproject.util.ChecksumType;
 import org.opencastproject.util.MimeTypes;
 import org.opencastproject.util.NotFoundException;
-import org.opencastproject.util.RequireUtil;
 import org.opencastproject.workspace.api.Workspace;

 import com.entwinemedia.fn.Fn;
@@ -107,7 +106,9 @@ public abstract class AbstractAssetManager implements AssetManager {
   /* ------------------------------------------------------------------------------------------------------------------ */

   @Override public Snapshot takeSnapshot(final String owner, final MediaPackage mp) {
-    RequireUtil.notEmpty(owner, "owner");
+    if (owner == null)
+      return takeSnapshot(mp);
+
     return handleException(new P1Lazy<Snapshot>() {
       @Override public Snapshot get1() {
         try {
```

That would change the behaviour to set the `DEFAULT_OWNER` instead of failing in case no owner was set before.

lkiesow commented:
Yes, that's basically what I meant.

And nothing prevents us from changing the behavior :)

mtneug committed:
MH-12758: Handle a null snapshot owner instead of failing

This changes the takeSnapshot logic to handle a null owner by first
checking if there is an owner for the last snapshot. If that is not the
case, the default owner is used.

mtneug commented:
`AbstractAssetManager` is now also updated.

staubesv commented:
@lkiesow & @mtneug : I've rated this as a release blocker as it doesn't seem to make much sense to release 4.2 before this one is merged.
